Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/jvm/main.kt

```diff
@@ -329,27 +329,42 @@ private fun processCLib(args: Array<String>, additionalArgs: Map<String, Any> =
         }
     }
 
+    val compiledFiles = compileSources(nativeLibsDir, tool, cinteropArguments)
+
     return when (stubIrOutput) {
         is StubIrDriver.Result.SourceCode -> {
             argsToCompiler(staticLibraries, libraryPaths)
         }
         is StubIrDriver.Result.Metadata -> {
-            val args = LibraryCreationArguments(
+            createInteropLibrary(
                     metadata = stubIrOutput.metadata,
-                    nativeBitcodePath = nativeOutputPath,
+                    nativeBitcodeFiles = compiledFiles + nativeOutputPath,
                     target = tool.target,
                     moduleName = moduleName,
                     outputPath = cinteropArguments.output,
                     manifest = def.manifestAddendProperties,
                     dependencies = allLibraryDependencies,
                     nopack = cinteropArguments.nopack
             )
-            createInteropLibrary(args)
             return null
         }
     }
 }
 
+private fun compileSources(
+        nativeLibsDir: String,
+        toolConfig: ToolConfig,
+        cinteropArguments: CInteropArguments
+): List<String> = cinteropArguments.compileSource.map { source ->
+    // Mangle path into file name to avoid collisions.
+    val mangledFileName = File(source).invariantSeparatorsPath.replace('/', '_')
+    val outputFileName = "$nativeLibsDir/${mangledFileName}.bc"
+    val compilerArgs = cinteropArguments.sourceCompileOptions.toTypedArray()
+    val compilerCmd = toolConfig.clang.clangCXX(*compilerArgs, source, "-emit-llvm", "-c", "-o", outputFileName)
```

Comment:
Probably won't work for C files as expected; `clang` should be used explicitly for these files.

Reply:
Maybe add separate CLI option that allows user to pick input language? Something like `-Xsource-language=c|c++` with `c++` by default.

Reply:
I guess the simplest solution right now is to keep it as is and don't recommend using C sources. Let's stick to this for now.